.github/actions/pmd/delta-scan.sh

```diff
@@ -0,0 +1,65 @@
+#!/usr/bin/env bash
+echo "=========================== Starting PMD Script ==========================="
+set -e
+
+# The git reference for the branch to merge the PR into.
+target_ref=$1
+# The git reference for the branch containing the PR changes.
+head_ref=$2
+# Set to "true" if an increase in PMD violations should cause the build to fail.
+fail_on_new_issues=$3
+
+# Requires pmd/pmd-github-action to have been executed already, as this will download PMD to this location.
+run_pmd="/opt/hostedtoolcache/pmd/${PMD_VERSION}/x64/pmd-bin-${PMD_VERSION}/bin/run.sh"
+
+# Create a temporary directory for storing files.
+tmp_dir=$(mktemp -d)
+
+# Create a list of the files changed by this PR.
+baseline_ref=$(git merge-base "${target_ref}" "${head_ref}")
+git diff --name-only ${baseline_ref} ${head_ref} > ${tmp_dir}/file-list.txt
+
+# Run PMD against the baseline commit.
+git checkout ${baseline_ref}
+for file in $(cat ${tmp_dir}/file-list.txt)
+do
+    if [[ -f ${file} ]]
+    then
+        echo ${file} > ${tmp_dir}/old-files.txt
+    fi
+done
+${run_pmd} pmd --cache ${tmp_dir}/pmd.cache --file-list ${tmp_dir}/old-files.txt -R ${ACTION_PATH}/pmd-ruleset.xml -r ${tmp_dir}/old_report.txt --fail-on-violation false
+old_issue_count=$(cat ${tmp_dir}/old_report.txt | wc -l)
+
+# Rerun PMD against the PR head commit.
+git checkout ${head_ref}
```

Comment:
I think you can simplify this script, when you need to checkout always use an external step `actions/checkout@v3` and then just call this (simplified) script with different parameters each time which just invoke pmd once

Reply:
I think that makes it harder to follow the behaviour really. As a really minor consideration it also makes the action less portable to other CI systems. If we extract the common code like this we would end up with something like:
* Run a script to find `baseline_ref` and to create `file-list.txt`
* Run checkout@v3
* Run the PMD script against the old commit, storing the results in `old_report.txt`
* Run checkout@v3
* Run the PMD script against the new commit, storing the results in `new_report.txt`
* Run a diff script to compare the violations and potentially fail the build.

I did consider extracting the common code (checkout, run PMD, count issues) into a bash function, but I didn't think it was worthwhile.